src/ws/session-utils.c

```diff
@@ -653,3 +565,134 @@ seal_memfd (FILE *memfd)
   const int seals = F_SEAL_SHRINK | F_SEAL_GROW | F_SEAL_WRITE;
   return fcntl (fileno (memfd), F_ADD_SEALS, seals) == 0;
 }
+
+static noreturn void
+__attribute__ ((format (printf, 1, 2)))
+abort_with_message (const char *format,
+                    ...)
+{
+  char buffer[1024];
+  va_list ap;
+
+  va_start (ap, format);
+  size_t length = vsnprintf (buffer, sizeof buffer, format, ap);
+  va_end (ap);
+
+  size_t ofs = 0;
+  while (ofs != length)
+    {
+      ssize_t r;
+      do
+        r = write (STDERR_FILENO, buffer + ofs, length - ofs);
+      while (r == -1 && errno == EINTR);
+
+      if (0 <= r && r <= length - ofs)
+        ofs += r;
+      else
+        break; /* something went wrong, but we can't deal with it */
+    }
+
+  abort ();
+}
+
+static void
+closefrom_via_proc (int first_fd)
+{
+  ssize_t bytes_read;
+
+  int proc_self_fd = open ("/proc/self/fd", O_RDONLY | O_DIRECTORY);
+  if (proc_self_fd == -1)
+    abort_with_message ("failed to close all file descriptors: open('/proc/self/fd'): %m");
+
+  do
+    {
+      char buffer[1024];
+      ssize_t ofs = 0;
+
+      /* getdents(2):
+       *
+       * Glibc does not provide a wrapper for these system calls; call them
+       * using syscall(2).  You will need to define the linux_dirent or
+       * linux_dirent64 structure yourself.
+       */
+      struct linux_dirent64 {
+        ino64_t        d_ino;    /* 64-bit inode number */
+        off64_t        d_off;    /* 64-bit offset to next structure */
+        unsigned short d_reclen; /* Size of this dirent */
+        unsigned char  d_type;   /* File type */
+        char           d_name[]; /* Filename (null-terminated) */
+      };
+
+      bytes_read = syscall (SYS_getdents64, proc_self_fd, buffer, sizeof buffer);
```

Comment:
This is an interesting point.  There are indeed two approaches to this problem, and maybe `O_CLOEXEC` is the better one.

On the other hand, leaking a fd from root (and one involved in authentication, no less) to the session is probably quite bad, so I can understand the desire to be explicit.

Reply:
Right, I understand the possibility of leaking an fd, but this is not unanimously better: This is so much code (and not "simple/obivously correct" code) that in the end the chance of introducing a bug with all this might be comparable to leaving an fd open.

Since Linux is lacking a closeall() thing and this is so excruciatingly hard to do in userspace, and solely relying on `O_CLOEXEC()` is not ideal either, here's another idea: Do use `O_CLOEXEC()`, and after fd remapping, try to close the next 5 fds -- if any of them does *not* result in `EBADF`, abort(), as we obviously messed up something. That's much much simpler than the /proc reading, abort_with_message() etc.,  and the chance of missing six or more fds should be practically zero.  AFAICS, pretty much the only API that could possibly open fds internally is the  kerberos negotiation, and we can check with strace how many it actually uses. But it's a bounded problem.

Reply:
Actually, `O_CLOEXEC` does not really help us here. If we have dangling open fds after the remapping, then they will still be open for that "check next 5 fds" loop -- after all, they will only close on the later execve(). So we don't really need to flag them with `O_CLOEXEC`, we really just need the loop to validate that we didn't introduce an fd leak. It seems fine to err on the side of safety and abort() if there is any unknown open fd.

Reply:
Let's chat about this out-of-band.  I disagree :)

Reply:
Summary from our conversation yesterday. The options are:
 - write a comprehensive unit test for this, to make sure it works on all architectures and does not leak
 - Be less generic, and after remapping just close the next 100 fds ≥ 4, and abort if any of them are still open (i. e. not `EBADF`) -- then we forgot to close one; for example, the forked process should close its PAM handle.
 - Use ulimit or similar to restrict cockpit-session to 20(ish) fds, and then just close them all.